packages/access-client/src/agent.js

```diff
@@ -409,6 +481,101 @@ export class Agent {
     }
   }
 
+  /**
+   * @param {Ucanto.DID<'key'>} space
+   * @param {Ucanto.Principal<Ucanto.DID<'mailto'>>} account
+   * @param {Ucanto.DID<'web'>} provider - e.g. 'did:web:staging.web3.storage'
+   */
+  async #addProvider(space, account, provider) {
+    return addProvider(this, space, account, provider)
+  }
+
+  /**
+   *
+   * @param {Ucanto.DID<'key'>} space
+   * @param {Ucanto.Principal<Ucanto.DID<'mailto'>>} account
+   */
+  async #delegateSpaceAccessToAccount(space, account) {
+    const issuerSaysAccountCanAdminSpace =
+      await createIssuerSaysAccountCanAdminSpace(
+        this.issuer,
+        space,
+        account,
+        undefined,
+        this.proofs([{ with: space, can: '*' }])
+      )
+    return this.invokeAndExecute(Access.delegate, {
+      audience: this.connection.id,
+      with: space,
+      expiration: Infinity,
+      nb: {
+        delegations: {
+          [issuerSaysAccountCanAdminSpace.cid.toString()]:
+            issuerSaysAccountCanAdminSpace.cid,
+        },
+      },
+      proofs: [
+        await createIssuerSaysDeviceCanAccessDelegateWithSpace(
+          this.issuer,
+          space,
+          this.issuer
+        ),
+        // must be embedded here because it's referenced by cid in .nb.delegations
+        issuerSaysAccountCanAdminSpace,
+      ],
+    })
+  }
+
+  /**
+   * Invokes voucher/redeem for the free tier, wait on the websocket for the voucher/claim and invokes it
+   *
+   * It also adds a full space delegation to the service in the voucher/claim invocation to allow for recovery
+   *
+   * @param {string} email
+   * @param {object} [opts]
+   * @param {AbortSignal} [opts.signal]
+   * @param {Ucanto.DID<'key'>} [opts.space]
+   * @param {Ucanto.DID<'web'>} [opts.provider] - provider to register - defaults to this.connection.id
+   */
+  async #newRegisterSpace(email, opts) {
+    const space = opts?.space || this.currentSpace()
+    const spaceMeta = space ? this.#data.spaces.get(space) : undefined
+    const provider =
+      opts?.provider ||
+      (() => {
+        const service = this.connection.id.did()
+        if (DIDValidator.match({ method: 'web' }).is(service)) {
+          // connection.id did is a valid provider value. Try using that.
+          return service
+        }
+        throw new Error(
+          `unable to determine provider to use to register space. Pass opts.provider`
```

Comment:
I think it would be a good idea to be more explicit and say that it tried to derive provider from the connection but it isn't did:web so perhaps there something wrong with connection or user could specify provider

Reply:
@gobengo